File: 542-01-matrix/01-matrix.cpp
class Solution {
public:
    vector<vector<int>> updateMatrix(vector<vector<int>>& mat) {
        int r = mat.size();
        int c = mat[0].size();
        vector<vector<int>> vis(r, vector<int> (c,0));
        vector<vector<int>> ans(r, vector<int> (c,0));
        queue<pair<pair<int,int>,int>> q;

        for (int i = 0; i < r; i++)
        {
            for (int j = 0; j < c; j++)
            {
                if (mat[i][j] == 0)
                {
                    q.push({{i,j},0});
                    vis[i][j] = 1;
                }
            }
        }

        vector<int> drow = {1,-1,0,0};
        vector<int> dcol = {0,0,1,-1};

        while (!q.empty())
        {
            int i = q.front().first.first;
            int j = q.front().first.second;
            int cdist = q.front().second;

            
            ans[i][j] = cdist;

            q.pop();

            for (int k = 0; k < 4; k++)
            {
                int ci = i + drow[k];
                int cj = j + dcol[k];


                if (ci >= 0 && ci < r && cj >= 0 && cj < c && vis[ci][cj] == 0)
                {
                    q.push({{ci,cj},cdist+1});
                    vis[ci][cj] = 1;
                }
            }
        }


        return ans;


    }
};

File: 542-01-matrix/01-matrix.py
from queue import Queue
class Solution:
    def updateMatrix(self, mat: List[List[int]]) -> List[List[int]]:
        r = len(mat)
        c = len(mat[0])

        vis = set()
        ans = [[0 for j in range(c)]for i in range(r)]
        q = Queue()

        for i in range(r):
            for j in range(c):
                if mat[i][j] == 0:
                    q.put((i,j,0))
                    vis.add((i,j))

        drow = [1,-1,0,0]
        dcol = [0,0,1,-1]
        
        while not q.empty():
            i,j,cdist = q.get()
            ans[i][j] = cdist

            for k in range(4):
                ci = i + drow[k]
                cj = j + dcol[k]

                if (ci >=0 and ci < r and cj >= 0 and cj < c and ((ci,cj) not in vis)):
                    q.put((ci,cj,cdist+1))
                    vis.add((ci,cj))
        
        return ans

        